src/par_iter/enumerate.rs

```diff
@@ -50,29 +51,50 @@ unsafe impl<M> ExactParallelIterator for Enumerate<M>
 impl<M> IndexedParallelIterator for Enumerate<M>
     where M: IndexedParallelIterator,
 {
-    type Producer = EnumerateProducer<M::Producer>;
+    fn with_producer<WP>(self, wp: WP) -> WP::Output
+        where WP: ProducerContinuation<Self::Item>
+    {
+        return self.base.with_producer(Continuation { wp: wp });
 
-    fn into_producer(self) -> (Self::Producer, <Self::Producer as Producer>::Shared) {
-        let (base, shared) = self.base.into_producer();
-        (EnumerateProducer { base: base, offset: 0 }, shared)
+        struct Continuation<WP> {
+            wp: WP,
+        }
+
+        impl<ITEM, WP> ProducerContinuation<ITEM> for Continuation<WP>
+            where WP: ProducerContinuation<(usize, ITEM)>
+        {
+            type Output = WP::Output;
+            fn with_producer<'produce, P>(self,
+                                          base: P,
+                                          shared: &'produce P::Shared)
+                                          -> WP::Output
+                where P: Producer<'produce, Item=ITEM>
+            {
+                let producer = EnumerateProducer { base: base,
+                                                   offset: 0,
+                                                   phantoms: PhantomType::new() };
+                self.wp.with_producer(producer, shared)
+            }
+        }
     }
 }
 
 ///////////////////////////////////////////////////////////////////////////
 // Producer implementation
 
-pub struct EnumerateProducer<M>
-    where M: Producer,
+pub struct EnumerateProducer<'p, P>
+    where P: Producer<'p>,
 {
-    base: M,
+    base: P,
     offset: usize,
+    phantoms: PhantomType<&'p ()>,
```

Comment:
As an aside, with this change, it would be possible to remove the `Shared` type from `Producer` and instead have each producer carry their own shared data internally as `&'p Foo` reference (copying it on each split). I started doing this but stopped for two reasons:

- It was kind of a pain and I'm sick of refactoring this code for a bit. :)
- It didn't seem to be making the code much simpler, somewhat surprisingly to me. Maybe a bit.
- It meant that each split had to copy more data, since instead of copying one shared reference, you copied N, one for each level. OTOH, to *access* the shared data, this version requires more dereferences. I have no idea which is better. Probably it doesn't make any difference at all.